### PR TITLE
upgrade to v0.9.16

### PR DESCRIPTION
**Pull request by zqhxuyuan**
**xcm related:**
- [x] **canonical reanchored location**: closes: https://github.com/AcalaNetwork/Acala/issues/1762
    - relates to: https://github.com/paritytech/polkadot/pull/4470 
    - I found it's difficult to setup config only to testcase, so I use `integration-tests` feature flag for now. 
- [x] **statemine asset prefix location**
    - relates to: https://github.com/paritytech/cumulus/pull/831
    - ~~also use `integration-tests` to testcase which make sure RMRK migration works.~~

**removal Default:**
- [x] MaxEncodedLen & without_storage_info
    - [x] stable-asset module should have this too, otherwise it'll compile error: https://github.com/nutsfinance/stable-asset/pull/14
- [x] ~~ParentIsPreset (not sure new cumulus release will include this yet)~~
- [x] CheckNonZeroSender: https://github.com/AcalaNetwork/Acala/issues/1662

client:
- [x] cumulus consensus refactor

runtime new pallet:
- [x] add pallet_preimage to pallet_scheduler
    - [x] migration: https://github.com/paritytech/substrate/pull/10356
- [x] ~~add pallet_child_bounties to pallet_bounties~~(do not add for now as kusama runtime not added)

tests:
- [x] ~~kusama native asset: https://github.com/AcalaNetwork/Acala/issues/1768~~
    - ~~current karura and sibling both use same runtime, so their native asset both is KAR, need `integration-tests` to manual adapt/fixed the para_id to 2000~~
- [x] test-service(version <= 0.9.15 has compile error of client type, so moved here)

---

upgrade guide:
- **canonical reanchored location**: before we upgrade, make sure bifrost already compatible. and current it's already done in bifrost side.
- **statemine asset prefix location** we don't need compatible because we use foreign_asset_id internal, but after statemine upgrade, we must change RMRK location on asset-registry from `(1, X2(1000, GenealIndex(8)))` to `(1, X3(1000, PalletInstance(50), GeneralIndex(8)))`

### CODE REVIEW

File: primitives/src/currency.rs [hunk 1]
```diff
@@ -205,6 +205,8 @@ create_currency_id! {
 		PHA("Phala Native Token", 12) = 170,
 		KINT("Kintsugi Native Token", 12) = 171,
 		KBTC("Kintsugi Wrapped BTC", 8) = 172,
+		RMRK("Rmrk", 10) = 173,
```
Comment (xlc) [inline primitives/src/currency.rs:208]: why add those?

File: runtime/acala/src/lib.rs [hunk 2]
```diff
@@ -921,6 +927,8 @@ impl orml_vesting::Config for Runtime {
 parameter_types! {
 	pub MaximumSchedulerWeight: Weight = Perbill::from_percent(10) * RuntimeBlockWeights::get().max_block;
 	pub const MaxScheduledPerBlock: u32 = 10;
+	// Retry a scheduled item every 10 blocks (1 minute) until the preimage exists.
+	pub const NoPreimagePostponement: Option<u32> = Some(10);
```
Comment (xlc) [inline runtime/acala/src/lib.rs:931]: 10 blocks is 2 mins for us. Use `MINUTES` to ensure it is 1 minute.
But I think we can make it 5 minutes here.

File: runtime/acala/src/lib.rs [hunk 3]
```diff
@@ -1912,8 +1959,32 @@ pub type SignedPayload = generic::SignedPayload<Call, SignedExtra>;
...
+	Runtime,
+	AllPalletsWithSystem,
+	SchedulerMigrationV3,
+>;
+
+// Migration for scheduler pallet to move from a plain Call to a CallOrHash.
+pub struct SchedulerMigrationV3;
+impl frame_support::traits::OnRuntimeUpgrade for SchedulerMigrationV3 {
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		Scheduler::migrate_v2_to_v3()
```
Comment (xlc) [inline runtime/acala/src/lib.rs:1975]: I think the current version we have is still V1

File: runtime/karura/Cargo.toml [hunk 4]
```diff
@@ -323,3 +326,5 @@ try-runtime = [
 	"module-homa-xcm/try-runtime",
 	"module-session-manager/try-runtime",
 ]
+
+integration-tests = []
```
Comment (xlc) [inline runtime/karura/Cargo.toml:330]: why is this needed?

File: runtime/karura/src/lib.rs [hunk 5]
```diff
@@ -920,6 +925,8 @@ impl orml_vesting::Config for Runtime {
 parameter_types! {
 	pub MaximumSchedulerWeight: Weight = Perbill::from_percent(10) * RuntimeBlockWeights::get().max_block;
 	pub const MaxScheduledPerBlock: u32 = 10;
+	// Retry a scheduled item every 10 blocks (1 minute) until the preimage exists.
+	pub const NoPreimagePostponement: Option<u32> = Some(10);
```
Comment (xlc) [inline runtime/karura/src/lib.rs:929]: make it 5 minutes

File: runtime/karura/src/lib.rs [hunk 6]
```diff
@@ -1545,7 +1571,46 @@ parameter_types! {
 	pub ForeignAssetUnitsPerSecond: u128 = kar_per_second();
 	pub KarPerSecondAsBased: u128 = kar_per_second();
 }
+#[cfg(feature = "integration-tests")]
+parameter_types! {
```
Comment (xlc) [inline runtime/karura/src/lib.rs:1575]: I think better to move those into another file

File: runtime/karura/src/lib.rs [hunk 7]
```diff
@@ -2011,8 +2131,32 @@ pub type SignedPayload = generic::SignedPayload<Call, SignedExtra>;
...
+	Runtime,
+	AllPalletsWithSystem,
+	SchedulerMigrationV3,
+>;
+
+// Migration for scheduler pallet to move from a plain Call to a CallOrHash.
+pub struct SchedulerMigrationV3;
+impl frame_support::traits::OnRuntimeUpgrade for SchedulerMigrationV3 {
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		Scheduler::migrate_v2_to_v3()
```
Comment (xlc) [inline runtime/karura/src/lib.rs:2147]: should be from v1 to v3

File: runtime/acala/src/lib.rs (hunk 3 above)
Comment (xlc) [reply runtime/acala/src/lib.rs:1975]: actually we are on v2 but the storage version in the pallet is still v1 for some reason, so we need to set the storage version before run the migration code

File: runtime/karura/Cargo.toml (hunk 4 above)
Comment (zqhxuyuan) [reply runtime/karura/Cargo.toml:330]: integration-tests use some special config only in testcase, but it's not required for production config. so I use `integration-tests` feature flag in runtime

File: runtime/acala/src/lib.rs (hunk 3 above)
Comment (zqhxuyuan) [reply runtime/acala/src/lib.rs]: change to `migrate_v1_to_v3 `

File: primitives/src/currency.rs (hunk 1 above)
Comment (zqhxuyuan) [reply primitives/src/currency.rs]: deleted now. when I test statemine prefix location break change. I use RMRK and RMRK_V2 as location migration.

File: runtime/acala/src/lib.rs (hunk 2 above)
Comment (zqhxuyuan) [reply runtime/acala/src/lib.rs]: change to 25 blocks, 25 * 12 = 300s = 5 min

File: runtime/karura/src/lib.rs (hunk 7 above)
Comment (zqhxuyuan) [reply runtime/karura/src/lib.rs]: done

**Comment by sander2** [comment]
Sorry to intrude on your PR, but I have a question. Do you know if the following change is required as well? I suspect that maybe polkadot/substrate will automatically canonicalize it, in which case it doesn't matter, but tbh I am not completely sure where the new `simplify` function gets called. I'm asking because we use some very similar runtime code.
```diff
fn native_currency_location(id: CurrencyId) -> MultiLocation {
-    MultiLocation::new(1, X2(Parachain(ParachainInfo::get().into()), GeneralKey(id.encode())))
+    MultiLocation::new(0, X1(GeneralKey(id.encode())))
}
```

**Comment by zqhxuyuan** [comment]
> 

don't need change `native_currency_location()`. as the new canonical reanchored only change the meaning of `to-reserve`/`non-reserve` case. i.e ParaB transfer token A to ParaA, from the point view of ParaA, the location is changed. But for the case of ParaA transfer token A to ParaB. we do not need to change `native_currency_location()` as xtokens will invoke reanchor, the for the point view of ParaB, the location not change, still got `(1, 2000, A)`.

File: primitives/src/signature.rs [hunk 8]
```diff
@@ -94,18 +94,24 @@ impl TryFrom<AcalaMultiSignature> for ecdsa::Signature {
...
-			(Self::Ed25519(ref sig), who) => sig.verify(msg, &ed25519::Public::from_slice(who.as_ref())),
-			(Self::Sr25519(ref sig), who) => sig.verify(msg, &sr25519::Public::from_slice(who.as_ref())),
+			(Self::Ed25519(ref sig), who) => match ed25519::Public::from_slice(who.as_ref()) {
+				Ok(signer) => sig.verify(msg, &signer),
+				Err(()) => false,
+			},
+			(Self::Sr25519(ref sig), who) => match sr25519::Public::from_slice(who.as_ref()) {
+				Ok(signer) => sig.verify(msg, &signer),
+				Err(()) => false,
+			},
```
Comment (shaunxw) [inline primitives/src/signature.rs:114]: use `map_or` to replace nested match

File: runtime/acala/src/lib.rs [hunk 9]
```diff
@@ -1705,6 +1731,21 @@ impl Convert<MultiLocation, Option<CurrencyId>> for CurrencyIdConvert {
 					_ => None,
 				}
 			}
+			// adapt for reanchor canonical location: https://github.com/paritytech/polkadot/pull/4470
+			MultiLocation {
+				parents: 0,
+				interior: X1(GeneralKey(key)),
+			} => {
+				let key = &key[..];
+				if let Ok(currency_id) = CurrencyId::decode(&mut &*key) {
```
Comment (shaunxw) [inline runtime/acala/src/lib.rs:1740]:
```rust
let currency_id = CurrencyId::decode(&mut &*key).ok()?;
```

File: runtime/acala/src/lib.rs [hunk 10]
```diff
@@ -933,6 +939,24 @@ impl pallet_scheduler::Config for Runtime {
 	type MaxScheduledPerBlock = MaxScheduledPerBlock;
 	type WeightInfo = ();
 	type OriginPrivilegeCmp = EqualPrivilegeOnly;
+	type PreimageProvider = Preimage;
+	type NoPreimagePostponement = NoPreimagePostponement;
+}
+
+parameter_types! {
+	pub const PreimageMaxSize: u32 = 4096 * 1024;
```
Comment (shaunxw) [inline runtime/acala/src/lib.rs:947]: Can we have comments about the actual size?

File: runtime/karura/src/lib.rs [hunk 11]
```diff
@@ -1804,10 +1837,29 @@ impl Convert<MultiLocation, Option<CurrencyId>> for CurrencyIdConvert {
...
 			} => Some(Token(PHA)),
+			// adapt for reanchor canonical location: https://github.com/paritytech/polkadot/pull/4470
+			MultiLocation {
+				parents: 0,
+				interior: X1(GeneralKey(key)),
+			} => match &key[..] {
+				#[cfg(feature = "integration-tests")]
+				parachains::bifrost::BNC_KEY => Some(Token(BNC)),
+				key => {
+					if let Ok(currency_id) = CurrencyId::decode(&mut &*key) {
```
Comment (shaunxw) [inline runtime/karura/src/lib.rs:1848]: let currency_id = CurrencyId::decode(&mut &*key).ok()?;

File: runtime/karura/src/lib.rs [hunk 12]
```diff
@@ -1804,10 +1837,29 @@ impl Convert<MultiLocation, Option<CurrencyId>> for CurrencyIdConvert {
...
+						match currency_id {
+							Token(KAR) | Token(KUSD) | Token(LKSM) => Some(currency_id),
```
Comment (shaunxw) [inline runtime/karura/src/lib.rs:1850]: do we need `TAI`

File: primitives/src/signature.rs [hunk 13]
```diff
@@ -94,18 +94,24 @@ impl TryFrom<AcalaMultiSignature> for ecdsa::Signature {
...
-impl Default for AcalaMultiSignature {
-	fn default() -> Self {
-		Self::Ed25519(Default::default())
-	}
-}
+// impl Default for AcalaMultiSignature {
+// 	fn default() -> Self {
+// 		Self::Ed25519(Default::default())
+// 	}
+// }
```
[the remark this answers is not in the document]
Comment (zqhxuyuan) [reply primitives/src/signature.rs:101]: there'll be error: the trait bound `sp_core::ed25519::Signature: std::default::Default` is not satisfied. as Default is removed everywhere. I'll delete this code, or change the parameter to some fixed value.

File: runtime/karura/src/lib.rs (hunk 12 above)
Comment (zqhxuyuan) [reply runtime/karura/src/lib.rs:1850]: good catch, I think `TAI` can be also consider as our parachain side token, then we should add it.

File: modules/evm/src/lib.rs [hunk 14]
```diff
@@ -1295,8 +1296,11 @@ impl<T: Config> Pallet<T> {
 		let digest = <frame_system::Pallet<T>>::digest();
 		let pre_runtime_digests = digest.logs.iter().filter_map(|d| d.as_pre_runtime());
 
-		let author = T::FindAuthor::find_author(pre_runtime_digests).unwrap_or_default();
-		T::AddressMapping::get_default_evm_address(&author)
+		if let Some(author) = T::FindAuthor::find_author(pre_runtime_digests) {
+			T::AddressMapping::get_default_evm_address(&author)
+		} else {
+			H160::default()
+		}
```
[the remark this answers is not in the document]
Comment (zqhxuyuan) [reply modules/evm/src/lib.rs:1303]: the `find_author()` return `Option(author)` value now, previous the return value is just `author`.

[the remark this answers is not in the document]
Comment (zjb0807) [reply modules/evm/src/lib.rs:1303]: LGTM.
`H160::default()` is better than `T::AddressMapping::get_default_evm_address(AccountId::default())`

File: ts-tests/tests/test-contract-methods.ts [hunk 15]
```diff
@@ -63,9 +63,9 @@ describeWithAcala("Acala RPC (Contract Methods)", (context) => {
 		expect((await contract.chainId()).toString()).to.eq('595');
 	});
 
-	it("should get correct environmental coinbase", async function () {
-		expect((await contract.coinbase()).toString()).to.not.eq('0x0000000000000000000000000000000000000000');
-	});
+	// it("should get correct environmental coinbase", async function () {
+	// 	expect((await contract.coinbase()).toString()).to.not.eq('0x0000000000000000000000000000000000000000');
```
Comment (zjb0807) [inline ts-tests/tests/test-contract-methods.ts:67]:
```suggestion
		expect((await contract.coinbase()).toString()).to.eq('0x0000000000000000000000000000000000000000');
```

File: modules/homa-lite/src/mock.rs [hunk 16]
```diff
@@ -105,6 +105,10 @@ impl XcmTransfer<AccountId, Balance, CurrencyId> for MockXcm {
 	}
 }
 impl InvertLocation for MockXcm {
+	fn ancestry() -> MultiLocation {
+		todo!()
```
[the remark this answers is not in the document]
Comment (zqhxuyuan) [reply modules/homa-lite/src/mock.rs]: add Parachain(2000) as Ancestry as this is a mock.

File: runtime/karura/src/lib.rs (hunk 12 above)
Comment (zqhxuyuan) [reply runtime/karura/src/lib.rs]: I checked mandala and acala runtime, `TAI` is only used in `ExistentialDeposits`, but not in `CurrencyIdConvert`, is it by design or a mistake? @xlc

File: runtime/karura/src/lib.rs [hunk 17]
```diff
@@ -132,6 +137,7 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
 	impl_version: 0,
 	apis: RUNTIME_API_VERSIONS,
 	transaction_version: 1,
+	state_version: 1,
```
Comment (sander2) [inline runtime/karura/src/lib.rs:140]: Just a heads-up: basti said on element that this has to be set to 0. I had it set to 1 initially as well, but it caused us to fail to produce blocks on a rococo-local setup

**Comment by sander2** [comment]
Another headsup: I found in kintsugi testing on rococo-local that we needed to canonicalize the chain-local assetids in the `PerSecond` Trader items

**Comment by zqhxuyuan** [comment]
> Another headsup: I found in kintsugi testing on rococo-local that we needed to canonicalize the chain-local assetids in the `PerSecond` Trader items

I'll revisit Trader config, as we're missing transfer native asset test, previous I add some code but delete them as I need to add `integration-tests` a few place. I think `KarPerSecondOfCanonicalLocation` should added to production case, not `integration-tests` case.

**Review by xlc — APPROVED** [review]
